R/g.analyse.perday.R

```diff
@@ -320,23 +324,50 @@ g.analyse.perday = function(ndays, firstmidnighti, time, nfeatures,
               # Comment out the following 10 lines if you want to include only the actual data
-              if (NRV != length(averageday[, (mi - 1)])) {
+              if (NRV < length(averageday[, (mi - 1)])) { # modified to avoid imputing time in the daylight saving days (25-h long, then they meet the condition NRV != nrow(averageday))
                 difference = NRV - length(averageday[, (mi - 1)])
                 if (di == 1) {
-                  varnum = c(averageday[1:abs(difference), (mi - 1)], varnum)
-                  # readjust anwi indices in case that varnum has been imputed
-                  if (max(anwi_t1) < length(varnum)) { # since GGIR always calculates full window, max(anwi_t1) should always equals length(varnum)
-                    anwi_t0 = anwi_t0 + abs(difference)
-                    anwi_t1 = anwi_t1 + abs(difference)
-                    # then, we reset the minimum anwi_t0 to 1 to consider the imputed varnum
-                    anwi_t0[which(anwi_t0 == min(anwi_t0))] = 1
+                  # varnum = c(averageday[1:abs(difference), (mi - 1)], varnum)
+                  varnum = c(rep(NA, abs(difference)), varnum) # 2022-11-1: if day <24h, then the total time analised is also <24h 
+                  # impute vari timestamp as well
+                  vari2 = matrix(NA, nrow = abs(difference), ncol = ncol(vari))
+                  colnames(vari2) = colnames(vari)
+                  ts1 = iso8601chartime2POSIX(vari[1,1], tz = desiredtz)
+                  if (is.na(ts1)) {
+                    ts1 = as.POSIXct(vari[1,1], tz = desiredtz)
                   }
+                  ts0 = ts1 - abs(difference)*ws3
+                  ts = seq(ts0, ts1, by = ws3)
+                  ts = ts[-length(ts)] #this timestamp is already in vari
+                  vari2[, 1] = POSIXtime2iso8601(ts, tz = desiredtz)
+                  vari2 = rbind(vari2, vari)
                 } else {
-                  a56 = length(averageday[,(mi - 1)]) - abs(difference)
-                  a57 = length(averageday[, (mi - 1)])
-                  varnum = c(varnum,averageday[a56:a57, (mi - 1)])
+                  varnum = c(varnum, rep(NA, abs(difference)))
+                  # impute vari timestamp as well
+                  vari2 = matrix(NA, nrow = abs(difference), ncol = ncol(vari))
+                  colnames(vari2) = colnames(vari)
+                  ts0 = iso8601chartime2POSIX(vari[nrow(vari),1], tz = desiredtz)
+                  if (is.na(ts0)) {
+                    ts0 = as.POSIXct(vari[nrow(vari),1], tz = desiredtz)
+                  }
+                  ts1 = ts0 + abs(difference)*ws3
+                  ts = seq(ts0, ts1, by = ws3)
+                  ts = ts[-1] #this timestamp is already in vari
+                  vari2[, 1] = POSIXtime2iso8601(ts, tz = desiredtz)
+                  vari2 = rbind(vari, vari2)
+                }
+                # readjust anwi indices in case that varnum has been imputed
```

Comment:
I see you moved `if (max(anwi_t1) < length(varnum)) {..}` down to be applied for both first and last day, while it used to be only applied for day 1. Only in day 1 we need to correct the indices for the difference because the time in the day is no longer the same, while in the last day this is not needed because the indicates are always correct because the start of the day is always midnight. See original [here](https://github.com/wadpac/GGIR/blob/master/R/g.analyse.perday.R#L330-L335).

Reply:
Right, I have replaced it where it was.